service/storage_service.cc

```diff
@@ -3043,26 +3042,68 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
 
     std::unordered_set<dht::token> tokens;
     if (!raft_topology_change_enabled()) {
-        tokens = get_tokens_for(replace_address);
-        if (tokens.empty()) {
+        auto tokens_opt = state->get_tokens();
+        if (!tokens_opt || tokens_opt->empty()) {
             throw std::runtime_error(::format("Could not find tokens for {} to replace", replace_address));
         }
+        tokens = std::move(*tokens_opt);
     }
 
     auto dc_rack = get_dc_rack_for(replace_address).value_or(locator::endpoint_dc_rack::default_location);
 
     if (!replace_host_id) {
         replace_host_id = _gossiper.get_host_id(replace_address);
     }
-    slogger.info("Host {}/{} is replacing {}/{}", get_token_metadata().get_my_id(), get_broadcast_address(), replace_host_id, replace_address);
-    co_await _gossiper.reset_endpoint_state_map();
 
-    co_return replacement_info {
+    auto ri = replacement_info {
         .tokens = std::move(tokens),
         .dc_rack = std::move(dc_rack),
         .host_id = std::move(replace_host_id),
         .address = replace_address,
     };
+
+    for (auto& hoep : parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace())) {
+        locator::host_id id;
+        gms::inet_address endpoint;
+        // Resolve both host_id and endpoint
+        if (hoep.has_endpoint()) {
+            endpoint = hoep.endpoint();
+        } else {
+            id = hoep.id();
+            auto res = _gossiper.get_nodes_with_host_id(id);
+            if (res.size() == 0) {
+                throw std::runtime_error(::format("Could not find ignored node with host_id {}", id));
+            } else if (res.size() > 1) {
+                throw std::runtime_error(::format("Found multiple nodes to ignore with host_id {}: {}", id, res));
+            }
+            endpoint = *res.begin();
+        }
+        auto esp = _gossiper.get_endpoint_state_ptr(endpoint);
+        if (!esp) {
+            throw std::runtime_error(::format("Ignore node {}/{} has no endpoint state", id, endpoint));
+        }
+        if (!id) {
+            id = esp->get_host_id();
+            if (!id) {
+                throw std::runtime_error(::format("Could not find host_id for ignored node {}", endpoint));
+            }
+        }
+        auto opt_dc_rack = esp->get_dc_rack();
+        tmptr->update_topology(id, opt_dc_rack);
+        tmptr->update_host_id(id, endpoint);
```

Comment:
`prepare_replacement_info` should not modify token_metadata. Previously it would return the obtained information to the caller, `join_token_ring`, who was responsible for updating token_metadata etc. Why not continue this pattern? And I still see that this is the case:
```cpp
        if (!raft_topology_change_enabled()) {
            bootstrap_tokens = std::move(ri->tokens);
            slogger.info("Replacing a node with {} IP address, my address={}, node being replaced={}",
                get_broadcast_address() == *replace_address ? "the same" : "a different",
                get_broadcast_address(), *replace_address);
            tmptr->update_topology(tmptr->get_my_id(), std::nullopt, locator::node::state::replacing);
            tmptr->update_topology(ri->host_id, std::move(ri->dc_rack), locator::node::state::being_replaced);
            co_await tmptr->update_normal_tokens(bootstrap_tokens, ri->host_id);
            tmptr->update_host_id(ri->host_id, *replace_address);
            replaced_host_id = ri->host_id;
        }
```

So first we update token_metadata here, and then again at the call site?

The more different places of updating state we have, the harder it is to understand the system.

Also, these updates should probably be disabled in raft-topology mode (like the updates I showed above), where responsibility of updating of token_metadata is shifted to group0 state machine.

Reply:
It's updating the state for the ignored nodes, not the replaced node.
Without that, with replace_ignore after full cluster restart the ignored nodes never go through handle_state_normal (since they are not considered normal token owners with a `NORMAL`  status in endpoint_state).

I'm ok with letting `join_token_ring` do that.
And I'll also test with consistent-topology-changes to make sure it's happening there too.

Reply:
> I'm ok with letting join_token_ring do that.

Yes please, let's put all the updates to token_metadata together in the code.

>  And I'll also test with consistent-topology-changes to make sure it's happening there too.

It should already be happening without this code, through `topology_state_load`, after joining node becomes a member of group 0. Once it performs the initial read_barrier it should include all nodes including the ignored ones in its token ring.

So we should be able to put this code in the `!raft_topology_change_enabled()` branch

Reply:
done in v10